### ISSUE TITLE
Question about training procedure of retriever

### ISSUE THREAD

**Issue by dayuyang1999** [texttron/tevatron#140]
Dear authors,

I was trying to reproduce repllama/repmistral and to understand the logic of the code.

# 1. Why use cross-entropy loss in contrastive learning?

in this file `/usa/dayu/Table_similarity/tevatron/src/tevatron/retriever/modeling/encoder.py`.

I have questions about the `forward` function.

Here cross-entropy loss function is used. The cross-entropy loss focuses on the probabilities assigned to the true classes of the samples. It doesn't directly account for the probabilities assigned to the incorrect (negative) classes in the loss calculation.

However, based on some [materials](https://lilianweng.github.io/posts/2021-05-31-contrastive/) I have read. Training a retriever needs to use contrastive loss, which pushes negatives away and pulls positive closer to the anchor. 

For example, in the triplet loss, it explicitly penalty on the negative distance between anchor and negatives. While Cross-entropy ignore this part.

![image](https://github.com/user-attachments/assets/d308e4bd-ac13-465f-a214-70ed86f7507f)

# 2. How  "in-batch" negative is considered? What if we have >1 positive?

Based on the code I saw In the `forward` function.

```python
            target = torch.arange(scores.size(0), device=scores.device, dtype=torch.long)
            target = target * (p_reps.size(0) // q_reps.size(0))
```

and the code under `/usa/dayu/Table_similarity/tevatron/src/tevatron/retriever/dataset.py`.

```python
        query = group['query']
        group_positives = group['positive_passages']
        group_negatives = group['negative_passages']

        formated_query = format_query(query, self.data_args.query_prefix)
        formated_passages = []

        if self.data_args.positive_passage_no_shuffle:
            pos_psg = group_positives[0]
        else:
            pos_psg = group_positives[(_hashed_seed + epoch) % len(group_positives)]
        
        formated_passages.append(format_passage(pos_psg['text'], pos_psg['title'], self.data_args.passage_prefix))

       ....

        for neg_psg in negs:
            formated_passages.append(format_passage(neg_psg['text'], neg_psg['title'], self.data_args.passage_prefix))

        return formated_query, formated_passages

```

If I understand correctly, assuming each query can have and only have 1 positive passage (must = 1), and `negative_size` numbers of negatives.

assuming `negative_size` = 2.

the positive index will be = [0, 3, 6, 9,...] . all others are negatives.

If I understand correctly, where does the `in-batch` negatives? And what if we have >1 positive passages?

Thanks!
